src/minidriver/minidriver.c

```diff
@@ -6381,14 +6381,26 @@ DWORD WINAPI CardGetProperty(__in PCARD_DATA pCardData,
 		if (dwFlags >= MD_MAX_PINS)
 			MD_FUNC_RETURN(pCardData, 1, SCARD_E_INVALID_PARAMETER);
 
-		if (!vs->pin_objs[dwFlags])
+		if (!dwFlags && !vs->pin_objs[dwFlags])
```

Comment:
At least in my case, when there is no PIN, it still returns `SCARD_E_INVALID_PARAMETER`, as this condition is met (`dwFlags == 0 && vs->pin_objs[dwFlags] == NULL`)

Possibly it should return "empty" PIN info for all `vs->pin_objs[dwFlags] == NULL` cases, as there may even be cards with ROLE_ADMIN and/or ROLE_USER (and not just ROLE_EVERYONE) without a PIN?

Reply:
OK, it should be:
`if (dwFlags && !vs->pin_objs[dwFlags])`
or to make it more readable:
`if (dwFlags != ROLE_EVERYONE)  && (vs->pin_objs[dwFlags] == NULL))`

As for none real pins `vs->pin_objs` is filled in from PKCS15 structures. 

Then  if `dwFlags == ROLE_EVERYONE` it will fall through to the  case ROLE_EVERYONE:'  and log a message and endup 
6510 with another log message and hexdump or *p

Will update and try again today.